### DISCUSSION TITLE
How to start a track at a particular time?

### DISCUSSION THREAD

**Discussion by maehem** [schreibfaul1/ESP32-audioI2S#333] [General]
In my project, I have five buttons, each select a different mp3 song.  Then the user presses a button, audio.connecttoFS loads a named file.  I want to start a the song at some particular time into the song.  So I've tried the sequence below but nothing ever happens. The track always starts at the beginning:

The snippet below is what I'm calling when the button is pressed.  The song plays fine, just never starts
the number of seconds I want (var 'pp')  into the song.
I commented them out but I show other hard coded things I have tried from similar methods.  Nothing ever works.

My setup is MP3 files (also tried WAV) on a Arduino ESP32 with a SD-card reader and an I2S amp. All the hardware  stuff is fine.

Any ideas?  Are WAV and MP3 not supported for skipping around a song?

```
    Serial.print( "Button " + btn );
    Serial.println("Pressed" );
    
    currentChannel = btn;

    char filePath[80];
    sprintf(filePath, "%d.wav", btn); // also tried mp3
    audio.connecttoFS(SD, filePath);
    
    audio.setAudioPlayPosition(pp); // Seconds into the song
    //audio.setFilePos(56);
    //audio.setTimeOffset(random(50));
    
```

**Comment by ClaudioFoscan** [comment]
Same Here ... 

Tried to use: 
`audio.connecttoFS(SD, filePath, _resumeFilePosition);`

or

`audio.setAudioPlayPosition(_resumeFilePosition); 
`

does not work with _resumeFilePosition in secounds ...
I use Mp3-Files ...
Any help to jump at any position in the mp3 File?

**Reply by ClaudioFoscan** [reply]
push ...

still not working, any help ?

**Reply by maehem** [reply]
I realized that the author was probably not going to fix it and did my solution with a Atmega328P.

**Reply by ClaudioFoscan** [reply]
So you mean, you wrote your own decoder code for an atmega328P?

anyway, so i'm the only one who wants to use this feature?

**Reply by olup** [reply]
I am also in need of this feature. Note that seek works for me if the file is already reading (I implemented+/- 10 seconds skips). But as of now I couldn't get my song to start at a certain moment. Maybe it needs to start for a bit before being able to seek.

**Comment by maehem** [comment]
I ended up using a circuit and code from AdaFruit as my starting project. Used the VS1053 Codec chip.

In my case I bought a cheap one on Amazon (and had to make a couple simple mods) but was able to get the code running and was able to make the MP3 seek work.  I can’t publish my final work though as it was for a client one-off.   Let me know if you try it and get stuck and I will try to copy paste snips of code.

If I can find some time in the next month, I’ll try to publish something more basic on GitHub that doesn’t violate my client’s use.

Mark



https://learn.adafruit.com/adafruit-vs1053-mp3-aac-ogg-midi-wav-play-and-record-codec-tutorial￼
Adafruit VS1053 MP3/AAC/Ogg/MIDI/WAV Codec Breakout Tutorial
learn.adafruit.com

**Reply by ClaudioFoscan** [reply]
Thanks for your reply Mark.

A different Setup but the ESP32 is not possible at the moment for me. so i have to figure out a solution with this library ...
i'll dig deeper into it ...
but let me know if you publish something, maybe a future project can use a different hw...

maybe schreibfaul1 is on it? or has some hints?

**Comment by schreibfaul1** [comment]
If I have an mp3 file and want to jump to a specific point in time (e.g. 45 seconds from the beginning), I need to know in advance the playing time of the audio block. The question is whether this can be done in a reasonable time with a µC like the ESP32.
So far I'm playing the file. If it is CBR the playing time can be determined exactly, with VBR the average of the first mp3 frames is used to estimate the total length.
How do you do that?

**Reply by ClaudioFoscan** [reply]
thanks, but it is not the problem to determine the exact position here, 
but to play from a specific time on ... 
as an example:
I play a audiobook. the player tells mit with     uint32_t getAudioFileDuration(); and uint32_t getAudioCurrentTime(); the time i'm at ...
if i want resume the book, i only need the time i was, +- a couple of seconds... normaly i would resume 10s befor that time...

if i use your Library in this way:

**audio.connecttoFS(SD, filePath, _resumeFilePosition);**
or
**audio.setAudioPlayPosition(_resumeFilePosition);**  

the track still plays from beginning as if the _resumeFilePosition is ignored ... 
so i assume you did not implement this.

is there a possibility to resume at the sepcific mp3-Frame/Buffer? 
can i use     **uint32_t getFilePos();** and **bool setFilePos(uint32_t pos);** to achive my needs? 
or do i understand here somethin completely wrong?

Thanks for any hints ...

**Comment by maehem** [comment]
This is from my very foggy memory, but the more I’ve thought about my project from last year and why I didn’t choose ESP32 was that I think I had determined that file seek is broken on Espressif's ESP32 libraries.  I had done a couple experiments to call and manipulate the seek command and it would never get the file position to there.  So it wouldn’t work in this audio library or any other.   This is from my fading memory of course.

Mark
